lib/connection_controller.go

```diff
@@ -330,6 +351,146 @@ func (cc *ConnectionController) connectValidators(activeValidatorsMap *collectio
+	for _, rn := range connectedOutboundRemoteNodes {
+		if excessiveOutboundRemoteNodes == 0 {
+			break
+		}
+		cc.rnManager.Disconnect(rn)
+		excessiveOutboundRemoteNodes--
+	}
+}
+
+// refreshNonValidatorInboundIndex is called periodically by the non-validator connector. It is responsible for
+// disconnecting excess inbound remote nodes.
+func (cc *ConnectionController) refreshNonValidatorInboundIndex() {
+	// First let's check if we have an excess number of inbound remote nodes. If we do, we'll disconnect some of them.
+	numConnectedInboundRemoteNodes := uint32(cc.rnManager.GetNonValidatorInboundIndex().Count())
+	excessiveInboundRemoteNodes := uint32(0)
+	if numConnectedInboundRemoteNodes > cc.targetNonValidatorInboundRemoteNodes {
+		excessiveInboundRemoteNodes = numConnectedInboundRemoteNodes - cc.targetNonValidatorInboundRemoteNodes
+	}
+	// Disconnect random inbound non-validators if we have too many of them.
+	inboundRemoteNodes := cc.rnManager.GetNonValidatorInboundIndex().GetAll()
+	for _, rn := range inboundRemoteNodes {
+		if excessiveInboundRemoteNodes == 0 {
+			break
+		}
+		cc.rnManager.Disconnect(rn)
+		excessiveInboundRemoteNodes--
+	}
+}
+
+func (cc *ConnectionController) connectNonValidators() {
+	// Only connect to addresses from the addrmgr if we don't specify --connect-ips. These addresses are *not* persistent,
+	// meaning if we disconnect from one we'll try a different one.
+	// TODO: We used this condition in the old code to prevent the node from connecting to non-connect-ips nodes.
+	// 	I'm not sure whether this is still necessary. I suppose the concern here was that the connect-ips nodes
+	//	should be prioritized over the addrmgr nodes, especially during syncing. However, I think we can achieve the
+	// 	same result by defining another flag, like a boolean --sync-from-persistent-peers-only, which could indicate that
+	//	we disregard non-persistent non-connect-ips nodes during syncing, if the flag is set to true.
+	// FIXME: This about uncommenting the below condition.
+	//if len(cc.connectIps) == 0 {
+	//	return
+	//}
+
+	numOutboundPeers := uint32(cc.rnManager.GetNonValidatorOutboundIndex().Count())
+
+	remainingOutboundPeers := uint32(0)
+	if numOutboundPeers < cc.targetNonValidatorOutboundRemoteNodes {
+		remainingOutboundPeers = cc.targetNonValidatorOutboundRemoteNodes - numOutboundPeers
+	}
+	for ii := uint32(0); ii < remainingOutboundPeers; ii++ {
+		addr := cc.getRandomUnconnectedAddress()
+		if addr == nil {
+			break
+		}
+		cc.AddrMgr.Attempt(addr)
+		// FIXME: error handle
+		cc.rnManager.CreateNonValidatorOutboundConnection(addr)
+	}
+}
+
+func (cc *ConnectionController) getRandomUnconnectedAddress() *wire.NetAddress {
+	for tries := 0; tries < 100; tries++ {
+		addr := cc.AddrMgr.GetAddress()
+		if addr == nil {
+			//glog.V(2).Infof("ConnectionManager.getRandomUnconnectedAddress: addr from GetAddressWithExclusions was nil")
+			break
```

Comment:
Why `break` and not `continue`?

Reply:
If the `GetAddress()` returns nil, the `AddrMgr` is empty. So we can break early.